color: add hex2color, color2hex, rgb2str, str2rgb

cgundogan opened the pull request:
I needed some conversions between colors given in hex code (string) and the `rgb_color_t` or `uint32_t` forms.
I thought sharing this might be a good idea.

authmillenon commented on sys/color/color.c:
```diff
@@ -127,3 +127,15 @@ void color_hsv2rgb(color_hsv_t *hsv, color_rgb_t *rgb)
+{
+    rgb->r = ((hex >> 16) & 0xFF);
+    rgb->g = ((hex >> 8) & 0xFF);
+    rgb->b = (hex & 0xFF);
```
Hacky, but more afficienent way could be

```C
uint32_t *ptr = (uint32_t *)rgb;
*ptr = hex << 8;
```

(please test this if you decide to use this. Might be, that I'm wrong about the shift.

cgundogan replied:
what if the struct has padding?

authmillenon commented on sys/color/color.c:
```diff
@@ -127,3 +127,15 @@ void color_hsv2rgb(color_hsv_t *hsv, color_rgb_t *rgb)
+
+void color_rgb2hex(color_rgb_t *rgb, uint32_t *hex)
+{
+    *hex = (rgb->r << 16) | (rgb->g << 8) | (rgb->b);
```
The reverse of above could be applied here

authmillenon commented:
Can you add a description about the intended use? It's not very clear to me.

cgundogan commented:
@authmillenon I added a description. I also amended `str2rgb` and `rgb2str` conversions, because I needed them for a demo. Furthermore, I introduced a very small unittest for checking the functionality of `rgb2str` and `str2rgb`

kaspar030 commented:
Now that you're touching this, why don't we save a colour as 32bit integer including alpha?

cgundogan commented:
@kaspar030 can you have a look at the last commit? Is that what you had in mind?

cgundogan commented on sys/include/color.h:
```diff
@@ -28,13 +29,17 @@ extern "C" {
-} color_rgb_t;
+typedef union {
+    uint32_t raw;
+    struct __attribute__((packed)) {
```
@kaspar030 do you think this attribute is superfluous here? I was not sure whether a 64bit platform e.g. might put padding into the argb struct (non-commit comment version)

authmillenon commented on tests/unittests/tests-color/tests-color.c:
```diff
@@ -0,0 +1,73 @@
+    const char *color_str = "F09A1D";
+    color_argb_t color;
+
+    TEST_ASSERT_EQUAL_INT(-1, color_str2rgb(color_str, &color));
```
What does this fail?

cgundogan replied:
it fails because the string does not start with an `#` sign

authmillenon replied:
Why is it required? Can't you save a few byte of ROM by just ignoring every non hex-char?

cgundogan replied:
yeah, you are right.

cgundogan commented:
@kaspar030 removed the packed attribute from the `argb_t` struct

cgundogan commented:
addressed @authmillenon's comments and squashed immediately.

cgundogan commented:
@authmillenon are you OK with the current state?

cgundogan commented:
@haukepetersen mentioned some concerns regarding the alpha value that this PR introduces. Could mention them here again?

haukepetersen commented:
@kaspar030: why should we want an alpha channel here? For simple stuff like LED stripes this is not needed - and so why don't we keep it as simple as possible?

cgundogan commented:
I removed the last commit from this PR, which introduced the alpha value.
The commit can be found in [1], in case the discussion keeps rolling.

@haukepetersen any objections to merge once travis is green?

[1] https://github.com/cgundogan/RIOT/tree/color_backup

haukepetersen commented:
One thing that might prove useful is to allow also lower-case letters when parsing strings (e.g. '#caffee'), but this we can add later, also.

ACK once you made Travis happy.

cgundogan commented:
thats possible currently, because the characters get transformed to upper case with `(str[0] &~ 0x20)`. Maybe I should add a testcase for that later

haukepetersen commented:
nice, overlooked that one... Merge at will once Travis agrees.

kaspar030 commented:
> @kaspar030: why should we want an alpha channel here? For simple stuff like LED stripes this is not needed - and so why don't we keep it as simple as possible?

Unless using multiple values or packed structs, it doesn't make a difference.

cgundogan commented:
added a unittest for the lower case str2rgb and fixed some travis complains

cgundogan commented:
travis is happy. @kaspar030 @haukepetersen I will merge this for now as this will be super handy for the demo. I will open a follow-up PR with the addition of the alpha value later so that the discussion can be continued.
